### DISCUSSION TITLE
HTTP Error 504: Gateway Time-out when compiling a .g4 file

### DISCUSSION THREAD

**Discussion by dblairchappell** [antlr/antlr4#4081] [Q&A]
This a strange one. Recently, when running running antlr4 on a grammar file, I've been getting the following error:

> File "C:\Python311\Lib\urllib\request.py", line 643, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 504: Gateway Time-out

Sometimes it happens and other times it doesn't. Why on earth antlr is trying to connect to anything, I'm not sure. At first I feared I might have a virus or something, but looking more closely it seems to be wanting to connect to maven:

>  File "C:\Python311\Lib\site-packages\antlr4_tool_runner.py", line 14, in latest_version
    with urlopen(f"https://search.maven.org/solrsearch/select?q=a:antlr4-master+g:org.antlr") as response:

So I guess antlr or python is trying to update something? Either way, it's very annoying and it's happening often enough to make compiling grammars impossible 50% of the time.

Does anyone have an idea of what is going on, and how to stop it?

Many thanks in advance

**Comment by kaby76** [comment]
You installed the [Antlr4 tools](https://github.com/antlr/antlr4-tools), and you are running Python on https://github.com/antlr/antlr4-tools/blob/master/antlr4_tool_runner.py. Prior to this tool, the instructions to run the Antlr4 tool were to (1) download an Antlr4 complete.jar file, (2) "install" Java, (3) set up an alias for running the Antlr tool using java. Unfortunately, it brought about a large number of problems because people did not know which .jar file (it has to be the "complete" .jar file), and an extraordinary number of issues in installing and running Java from the command line. This Python tool removes the problems of the "old way". You can go back to explicitly downloading the .jar and calling java if you want. But, if the Python script is timing out 50% of the times you use it, then it's likely the Python tool should be improved by outputting that it's in a spin loop trying on a network failure.

I implemented a tool for C# builds ([Antlr4BuildTasks](https://github.com/kaby76/Antlr4BuildTasks)) that does much of the same, namely query the network and get what is needed.
